### ISSUE TITLE
onedriver deletes edited file remotely and cannot upload a new one

### ISSUE THREAD

**Issue by 0anton** [jstaf/onedriver#74]
Hi Jeff,

I tried now the latest version of onedriver [1] and noticed a strange behavior. I edited little text file (2 lines) and saved it after adding one line. onedriver effectively deleted it remotely, while preserving local version. Here are my logs:

Initial edit (file was still on onedrive):

```
DEBU[2020-04-19T09:31:36]inode.go:784 003297:fs.(*Inode).Open() Opening file for I/O.                         id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
INFO[2020-04-19T09:31:36]inode.go:825 003297:fs.(*Inode).Open() Found content in cache.                       id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
DEBU[2020-04-19T09:31:36]inode.go:612 000042:fs.(*Inode).Create()                                               id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L mode=100600 name=".How to squash commits to keep repo concise and cle.txt.swp" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:31:36]inode.go:612 000042:fs.(*Inode).Create()                                               id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L mode=100600 name=".How to squash commits to keep repo concise and cle.txt.swx" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:31:36]inode.go:461 002474:fs.(*Inode).Flush()                                               id=local-jARQT2RVhVrt13P6i5xC path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swx"
DEBU[2020-04-19T09:31:36]inode.go:428 002474:fs.(*Inode).Fsync()                                               id=local-jARQT2RVhVrt13P6i5xC path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swx"
DEBU[2020-04-19T09:31:36]inode.go:660 003297:fs.(*Inode).Unlink() Unlinking inode.                              id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L name=".How to squash commits to keep repo concise and cle.txt.swx" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:31:36]inode.go:461 002474:fs.(*Inode).Flush()                                               id=local-PYK7vmub2X8uXIu6FNcJ path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swp"
DEBU[2020-04-19T09:31:36]inode.go:428 002474:fs.(*Inode).Fsync()                                               id=local-PYK7vmub2X8uXIu6FNcJ path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swp"
DEBU[2020-04-19T09:31:36]inode.go:660 000042:fs.(*Inode).Unlink() Unlinking inode.                              id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L name=".How to squash commits to keep repo concise and cle.txt.swp" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:31:36]inode.go:612 000042:fs.(*Inode).Create()                                               id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L mode=100600 name=".How to squash commits to keep repo concise and cle.txt.swp" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:31:36]inode.go:461 000042:fs.(*Inode).Flush()                                               id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
DEBU[2020-04-19T09:31:36]inode.go:428 000042:fs.(*Inode).Fsync()                                               id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
DEBU[2020-04-19T09:31:36]inode.go:784 002474:fs.(*Inode).Open() Opening file for I/O.                         id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
INFO[2020-04-19T09:31:36]inode.go:825 002474:fs.(*Inode).Open() Found content in cache.                       id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
DEBU[2020-04-19T09:31:36]inode.go:461 003297:fs.(*Inode).Flush()                                               id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
DEBU[2020-04-19T09:31:36]inode.go:428 003297:fs.(*Inode).Fsync()                                               id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 path="/SAPDocuments/_work_notes/How to squash commits to keep repo concise and cle.txt"
DEBU[2020-04-19T09:31:50]delta.go:21 000022:fs.(*Cache).DeltaLoop() Fetching deltas from server.                 
INFO[2020-04-19T09:31:50]delta.go:44 000022:fs.(*Cache).DeltaLoop() Fetched 3 deltas.                            
DEBU[2020-04-19T09:31:50]delta.go:122 000022:fs.(*Cache).applyDelta() Applying delta                                id=01USGCVP6NSQBTPVVNLFE2VP6CCTQQDNKF name=
INFO[2020-04-19T09:31:50]delta.go:149 000022:fs.(*Cache).applyDelta() Applying server-side deletion of item.        delta=delete id=01USGCVP6NSQBTPVVNLFE2VP6CCTQQDNKF name=
DEBU[2020-04-19T09:31:50]delta.go:122 000022:fs.(*Cache).applyDelta() Applying delta                                id=01USGCVP56Y2GOVW7725BZO354PWSELRRZ name=root
DEBU[2020-04-19T09:31:50]delta.go:122 000022:fs.(*Cache).applyDelta() Applying delta                                id=01USGCVP72RGN57VEFC5HKCE6QECKAZZAA name=
INFO[2020-04-19T09:31:50]delta.go:149 000022:fs.(*Cache).applyDelta() Applying server-side deletion of item.        delta=delete id=01USGCVP72RGN57VEFC5HKCE6QECKAZZAA name=
INFO[2020-04-19T09:31:50]cache.go:558 000022:fs.(*Cache).SerializeAll() Serializing cache metadata to disk.          
DEBU[2020-04-19T09:31:57]inode.go:461 000042:fs.(*Inode).Flush()                                               id=local-rL5Fc3GcuHC03kaunATU path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swp"
DEBU[2020-04-19T09:31:57]inode.go:428 000042:fs.(*Inode).Fsync()                                               id=local-rL5Fc3GcuHC03kaunATU path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swp"
DEBU[2020-04-19T09:31:57]inode.go:660 003297:fs.(*Inode).Unlink() Unlinking inode.                              id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L name=".How to squash commits to keep repo concise and cle.txt.swp" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:31:59]upload_session.go:191 004515:fs.(*UploadSession).Upload() Uploading file.                               id=01USGCVP6RPID7ILVPOJFIPOFYRMEMDLYV
ERRO[2020-04-19T09:31:59]upload_session.go:216 004515:fs.(*UploadSession).Upload() Error during small file upload.               err="HTTP 404 - itemNotFound: The resource could not be found." id=01USGCVP6RPID7ILVPOJFIPOFYRMEMDLYV response=
ERRO[2020-04-19T09:32:01]upload_manager.go:46 000055:fs.(*UploadManager).uploadLoop() Upload failed.                                id=01USGCVP6RPID7ILVPOJFIPOFYRMEMDLYV

```

Second attempt to upload file after another local fails too:

```
DEBU[2020-04-19T09:26:24]inode.go:660 000027:fs.(*Inode).Unlink() Unlinking inode.                              id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L name="How to squash commits to keep repo concise and cle.txt~" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:26:24]inode.go:461 000028:fs.(*Inode).Flush()                                               id=local-VbhV3vC5AWX39IVUWSP2 path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swp"
DEBU[2020-04-19T09:26:24]inode.go:428 000028:fs.(*Inode).Fsync()                                               id=local-VbhV3vC5AWX39IVUWSP2 path="/SAPDocuments/_work_notes/.How to squash commits to keep repo concise and cle.txt.swp"
DEBU[2020-04-19T09:26:25]inode.go:660 000042:fs.(*Inode).Unlink() Unlinking inode.                              id=01USGCVP733ZIJ4JG3B5EL4HK7V5BYTD5L name=".How to squash commits to keep repo concise and cle.txt.swp" path=/SAPDocuments/_work_notes
DEBU[2020-04-19T09:26:25]upload_session.go:191 000475:fs.(*UploadSession).Upload() Uploading file.                               id=01USGCVP4CXU2CXPSNE5AIA7IH4XD6HXPP
DEBU[2020-04-19T09:26:25]upload_session.go:191 000474:fs.(*UploadSession).Upload() Uploading file.                               id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424
ERRO[2020-04-19T09:26:25]upload_session.go:216 000475:fs.(*UploadSession).Upload() Error during small file upload.               err="HTTP 404 - itemNotFound: The resource could not be found." id=01USGCVP4CXU2CXPSNE5AIA7IH4XD6HXPP response=
ERRO[2020-04-19T09:26:27]upload_manager.go:46 000055:fs.(*UploadManager).uploadLoop() Upload failed.                                id=01USGCVP4CXU2CXPSNE5AIA7IH4XD6HXPP
DEBU[2020-04-19T09:26:35]delta.go:21 000022:fs.(*Cache).DeltaLoop() Fetching deltas from server.                 
INFO[2020-04-19T09:26:35]delta.go:44 000022:fs.(*Cache).DeltaLoop() Fetched 6 deltas.                            
DEBU[2020-04-19T09:26:35]delta.go:122 000022:fs.(*Cache).applyDelta() Applying delta                                id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 name="How to squash commits to keep repo concise and cle.txt"
INFO[2020-04-19T09:26:35]delta.go:210 000022:fs.(*Cache).applyDelta() Overwriting local item, no local changes to preserve.  delta=overwrite id=01USGCVPY3EP3PJ76SZFGYGAPVBRBGE424 name="How to squash commits to keep repo concise and cle.txt"
DEBU[2020-04-19T09:26:35]delta.go:122 000022:fs.(*Cache).applyDelta() Applying delta                                id=01USGCVP56Y2GOVW7725BZO354PWSELRRZ name=root
DEBU[2020-04-19T09:26:35]delta.go:122 000022:fs.(*Cache).applyDelta() Applying delta                                id=01USGCVP2JU76MZZJGCZG36HLCOEJG3O2J name=SAPDocuments
INFO[2020-04-19T09:26:35]delta.go:225 000022:fs.(*Cache).applyDelta() Skipping, no changes relative to local state.  delta=skip id=01USGCVP2JU76MZZJGCZG36HLCOEJG3O2J name=SAPDocuments
```

[1] https://github.com/jstaf/onedriver/commit/49991e7d247b25cf0cc15ea3cf1dc02be07d5be4

**Comment by jstaf** [comment]
I'll take a look and see if I can get this fixed. :slightly_smiling_face:

**Comment by jstaf** [comment]
I haven't been able to reproduce the error, but I did spend some time this weekend rewriting how uploads are handled and making the process significantly more robust. Can you try rebuilding onedriver from the master branch and let me know if this solves the issue?

The `HTTP 404 - itemNotFound: The resource could not be found.` error during uploads may not be the actual error, as I found that sometimes happened for very short lived files when editing with vim (vim creates a `.filename.txt.swx`, `.filename.txt.swp`, `.filename.txt~`, and `4913` files for very brief moments of time while opening and writing files. Regardless, if the 404 still happens, there's actually retries for it now.

**Comment by jstaf** [comment]
This should now be solved in version 0.9.0, but if not, please reopen this issue and let me know.
